File: LongestSubStringWithoutRepeatingCharacters.java
//time-O(n)
//space-O(1)
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        if(s==null || s.length()==0) return max;
        Map<Character, Integer> map = new HashMap<>();
        int p1 = 0;

        for(int i=0; i<s.length(); i++){
            char ch = s.charAt(i);
            if(map.containsKey(ch)){
                p1 = Math.max(p1, map.get(ch)+1);
            }
            map.put(ch, i);
            max = Math.max(max, i-p1+1);
        }

        return max;
    }
}

File: Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Hash Set Approach
        # Time O(2n)
        # Space O(k)
        # hashset = set()
        # maxlen = 0
        # currlen = 0
        # p1 = 0
        # for p2 in range(len(s)):
        #     if s[p2] not in hashset:
        #         hashset.add(s[p2])
        #         currlen += 1
        #     else:

        #         while s[p1] != s[p2]:
        #             hashset.remove(s[p1])
        #             p1 += 1
        #             currlen -= 1
        #         p1 += 1
        #         # currlen -= 1
        #     maxlen = max(maxlen, currlen)
        #     # print(currlen)
        #     # print(p1,p2)
        # return maxlen

        # Hash Map approach
        # Time O(n)
        # Space O(k)
        hmap = {}
        maxlen = 0
        slow = 0
        for i in range(len(s)):
            ch = s[i]
            if ch in hmap:
                if hmap[ch] >= slow:
                    slow = hmap[ch] + 1

            hmap[ch] = i
            maxlen = max(maxlen, i - slow + 1)

        # print(hmap)
        return maxlen




